image/tarexport/loader.go

```diff
@@ -0,0 +1,136 @@
+package tarexport
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/reference"
+	"github.com/docker/docker/layer"
+	"github.com/docker/docker/pkg/progress"
+	ociv1 "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+type loader interface {
+	Load() error
+}
+
+type dockerLoader struct {
+	manifests      []manifestItem
+	tr             *tarexporter
+	tmpDir         string
+	outStream      io.Writer
+	progressOutput progress.Output
+}
+
+func (dl *dockerLoader) Load() error {
+	return dl.tr.loadHelper(dl.manifests, dl.outStream, dl.progressOutput, dl.tmpDir)
+}
+
+type v0TarLoader struct {
+	tr             *tarexporter
+	tmpDir         string
+	outStream      io.Writer
+	progressOutput progress.Output
+}
+
+func (dll *v0TarLoader) Load() error {
+	return dll.tr.legacyLoad(dll.tmpDir, dll.outStream, dll.progressOutput)
+}
+
+type ociLoader struct {
+	tr             *tarexporter
+	tmpDir         string
+	outStream      io.Writer
+	progressOutput progress.Output
+}
+
+func (ol *ociLoader) Load() error {
+	if ol.tr.name != "" && len(ol.tr.refs) != 0 {
+		return errors.New("cannot load with both name and refs")
+	}
+	if ol.tr.name == "" && len(ol.tr.refs) == 0 {
+		return errors.New("no OCI name mapping provided")
+	}
+
+	// FIXME(runcom): validate and check version of "oci-layout" file
+
+	manifests := []manifestItem{}
+	indexJSON, err := os.Open(filepath.Join(ol.tmpDir, "index.json"))
+	if err != nil {
+		return err
+	}
+	defer indexJSON.Close()
+	index := ociv1.Index{}
+	if err := json.NewDecoder(indexJSON).Decode(&index); err != nil {
+		return err
+	}
+	for _, md := range index.Manifests {
+		if md.MediaType != ociv1.MediaTypeImageManifest {
+			continue
+		}
+		d := md.Digest
+		manifestPath := filepath.Join(ol.tmpDir, "blobs", d.Algorithm().String(), d.Hex())
+		f, err := os.Open(manifestPath)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+		man := ociv1.Manifest{}
+		if err := json.NewDecoder(f).Decode(&man); err != nil {
+			return err
+		}
+		layers := make([]string, len(man.Layers))
+		foreignSrc := make(map[layer.DiffID]distribution.Descriptor)
+		for i, l := range man.Layers {
+			layerDigest := l.Digest
+			layers[i] = filepath.Join("blobs", layerDigest.Algorithm().String(), layerDigest.Hex())
+			if len(l.URLs) != 0 {
+				foreignSrc[layer.DiffID(layerDigest)] = distribution.Descriptor{
+					Digest: l.Digest,
+					Size:   l.Size,
+					URLs:   l.URLs,
+				}
+			}
+		}
+		var tag string
+		refName, ok := md.Annotations[ociv1.AnnotationRefName]
```

Comment:
Discussing if we should make this customisable, so that a different annotation can be used as name instead of the standard one.

Reply:
why would we ever differ from what OCI states by introducing/reading a custom docker annotation for an OCI image?

Reply:
What's current status?

Reply:
I'm with @runcom -- this doesn't make any sense and would just lead to fragmentation among OCI image users. The main proponent of the `org.opencontainers.ref.name` setup was @stevvooe -- and I'm sure he had considered integration into Docker as part of the design.

Reply:
@cyphar I have always been against having naming in OCI, as it causes these kinds of problems. I've only been around to contain the damage. Please stop harassing people with nonsense.

There are two considerations to make with `ref.name`:

1. It should express a reference to the granularity of a tag.
2. It is only valid on index.json to express tags with an archive.

The problem is that the contents of OCI `ref.name` may differ from what can be supported by docker. To avoid conflict between Docker and whatever OCI decided to put in `ref.name`, one can define a vendor specific field that has a more restrictive format that is compatible with docker.

If you want compatibility, you either need to lock `ref.name` down to what docker supports or support vendor-specific tags.

Reply:
> Please stop harassing people with nonsense.
Woah, simmer. I didn't see anything harassing.

Reply:
> I have always been against having naming in OCI, as it causes these kinds of problems. I've only been around to contain the damage. Please stop harassing people with nonsense.

It wasn't my intention to harass you at all, sorry if I made you feel that way. The reason why I mentioned you is because you did quite a bit of the work on `ref.name` as well as the main person I remember being a major voice in the `ref.name` discussions was you (hence as far as I'm aware you are the resident expert on it). I wanted to hear your opinion on it.

For the record, I agree with you that naming is something that generally should've been ironed out a bit more (though I don't really agree it should've been excluded entirely, as that would have caused usability issues in my humble opinion) -- but we now have `ref.name` and I had always assumed the background of the design had considered how containerd/Docker would interact with it. If I'm mistaken in that assumption, please let me know.

> If you want compatibility, you either need to lock ref.name down to what docker supports or support vendor-specific tags.

I don't mind supporting vendor specific tags, but if the idea of `docker save --format=oci` is to be compatible with other tools then I think compatibility of the "entrypoint" to the image with other tools needs to be considered. Every OCI image tool (`umoci`, `skopeo`, and `containerd` as far as I know) all use `ref.name` -- so if Docker starts outputting things not using `ref.name` then it will have compatibility issues (if you recall this was part of one of the longer discussions we had last year).

I understand the concern with Docker putting stuff in `ref.name` that is incompatible -- maybe we should have some way of telling Docker to remap the image names for OCI images so that they are compatible with `ref.name`? Or maybe Docker should output both `ref.name` (in an encoded format to avoid the character set issue -- so that at least it's usable out-of-the-box with other tools) and also a "canonical" `docker.ref.name` which is vendor-specific and thus tools can grow support for it organically.

Reply:
@cyphar The concern is really in the other direction. If other tools place stuff in `ref.name` that is incompatible, docker won't be able to import those without munging the names. The scope of what can be done in `ref.name` is greater than the scope of what docker can or will support.

While it would have been nice to specify OCI `ref.name`, there were too many requests for changes to make it work without changing docker. I'll also concede that the grammar is pretty bonkers, so specifying it formally would have been problematic.

I would say that as long as people adhere to what docker supports, there won't be a problem. If there are systems that want to use their own naming and be compatible with docker, the vendor specific `ref.name` would be an avenue of support.

As far as naming containers in the future goes, you can see some of my latest opinions in https://github.com/containerd/containerd/blob/master/reference/reference.go#L39. This creates a docker-compatible super-set based on schema-less URIs. The host becomes a namespace or authority and registry lookup is done out of band. This is what containerd implements but it defers interpretation of the url up to the actual subsystem that locates the image, rather than relying on microformat details.

Reply:
What if `docker load --format=oci` took a flag that told it what the in-Docker image name should be and what `ref.name` to source it from should be -- so that at the very least you didn't have to add Docker-specific tags in order to use the OCI import?

Speaking personally, I would prefer having to manually do `docker load --format=oci --output cyphar/blah:foo --input some_oci_image:foo` rather than having to modify every image I generate so that they can be imported into Docker.

And there is the obvious question of if we already have this issue with just `docker load` -- how will this problem shake out when (if?) Docker supports OCI images through `docker pull`? I will be honest that I'm not familiar how the OCI distribution-spec has changed since it was copied from the Docker one -- but I imagine that `ref.name` is something that we would want to have some sort of unity over? Or will the naming on the distribution-spec level be entirely separate to the naming on the image-spec level?

(As for naming, I think that there should be some even more drastic discussions about naming but I think those are *way* out of scope here.)

Reply:
@stevvooe 

What if `docker load --format=oci` took a flag that told it what the in-Docker image name should be and what `ref.name` to source it from should be -- so that at the very least you didn't have to add Docker-specific tags in order to use the OCI import?

Speaking personally, I would prefer having to manually do `docker load --format=oci --output cyphar/blah:foo --input some_oci_image:foo` rather than having to modify every image I generate so that they can be imported into Docker.

And there is the obvious question of if we already have this issue with just `docker load` -- how will this problem shake out when (if?) Docker supports OCI images through `docker pull`? I will be honest that I'm not familiar how the OCI distribution-spec has changed since it was copied from the Docker one -- but I imagine that `ref.name` is something that we would want to have some sort of unity over? Or will the naming on the distribution-spec level be entirely separate to the naming on the image-spec level?

(As for naming, I think that there should be some even more drastic discussions about naming but I think those are *way* out of scope here.)